"Virtual" pages prototype (#1175)
withastro/starlight: docs/src/content/docs/guides/pages.mdx

---
title: Pages
description: Learn how to create and manage your documentation site’s pages with Starlight.
sidebar:
  order: 1
---

Starlight generates your site’s HTML pages based on your content, with flexible options provided via Markdown frontmatter.

> In addition, Starlight projects have full access to Astro’s powerful page generation tools.

Reviewer: I was thinking we could have a local content and remote content section in the new "pages" page. My assumption is that Virtual pages main use case is tied with fetching content and displaying it in an Starlight layout.

> In addition, Starlight projects have full access to Astro’s powerful page generation tools.

Reviewer 2: Yeah the use cases are:
- fetching content from an external source
- generating content from non-Markdown local data
- plugins injecting a route that do the above
- probably something else I haven’t though of yet

This guide shows how page generation works in Starlight.

> ## Content pages

Reviewer 2: Yeah the naming was hard here. My thinking was that these are in `src/content/`, but yeah it’s not perfect. They’d probably _both_ be “docs pages”, but maybe something like “autogenerated” could work… 🤔

### File formats

Starlight supports authoring content in Markdown and MDX with no configuration required.
You can add support for Markdoc by installing the experimental [Astro Markdoc integration](https://docs.astro.build/en/guides/integrations-guide/markdoc/).

### Add pages

Add new pages to your site by creating `.md` or `.mdx` files in `src/content/docs/`.
Use sub-folders to organize your files and to create multiple path segments.

For example, the following file structure will generate pages at `example.com/hello-world` and `example.com/reference/faq`:

import FileTree from '~/components/file-tree.astro';

<FileTree>

- src/
  - content/
    - docs/
      - hello-world.md
      - reference/
        - faq.md

</FileTree>

### Type-safe frontmatter

All Starlight pages share a customizable [common set of frontmatter properties](/reference/frontmatter/) to control how the page appears:

```md
---
title: Hello, World!
description: This is a page in my Starlight-powered site
---
```

If you forget anything important, Starlight will let you know.

## Custom pages

> For advanced use cases, you can create custom pages in Astro’s `src/pages/` directory.

Reviewer 3: ```suggestion
For advanced use cases, you can create custom pages by adding Astro’s reserved `src/pages/` directory.
```
reserved? special? But in any case, the point I think we really need to get across is from a default Starlight site, you'll have to MAKE this folder yourself!

Reviewer 2: Ah, right. That’s true that it might not exist. The suggestion makes it sound like just adding a directory creates pages — will have a thing about wording.

This is helpful if you need to build pages with a completely custom layout or generate a page from an alternative data source.

> The `src/pages/` directory uses file-based routing and includes support for `.astro` files.

Reviewer 3: Just checking: ONLY `.astro` pages? Not HTML, MDX etc. Because that's what this implies.

Reviewer 2: No, all formats Astro supports are supported, this is what the “includes” was intended to imply. I didn’t want to duplicate the full list from Astro docs when people should use that as the source of truth.

> The `src/pages/` directory uses file-based routing and includes support for `.astro` files.

Reviewer 3: ```suggestion
The `src/pages/` directory uses [Astro's file-based routing](https://docs.astro.build/en/basics/astro-pages/#file-based-routing) and includes support for `.astro` files.
```

	<CustomComponent />
</StarlightPage>
```

#### Props

The `<StarlightPage />` component accepts the following props.

##### `frontmatter` (required)

**type:** `StarlightPageFrontmatter`

Set the [frontmatter properties](/reference/frontmatter/) for this page, similar to frontmatter in Markdown pages.
The [`title`](/reference/frontmatter/#title-required) property is required and all other properties are optional.

The following properties differ from Markdown frontmatter:

- The [`slug`](/reference/frontmatter/#slug) property is not supported and is automatically set based on the custom page’s URL.
- The [`editUrl`](/reference/frontmatter/#editurl) option requires a URL to display an edit link.
- The [`sidebar`](/reference/frontmatter/#sidebar) property is not supported. In Markdown frontmatter, this option allows customization of [autogenerated link groups](/reference/configuration/#sidebar), which is not applicable to pages using the `<StarlightPage />` component.

##### `sidebar`

**type:** `SidebarEntry[] | undefined`  
**default:** the sidebar generated based on the [global `sidebar` config](/reference/configuration/#sidebar)

> Provide a custom site navigation sidebar for this page.

Reviewer 3: Just pointing out that the structure of all of these definitions is an action the user might want to take, vs e.g.

"A property indicating whether Starlight's default global sidebar is used" or "The sidebar entry to use"

Also, I have no idea how I would set this. I think an example here would be helpful to show what that looks like.

Reviewer 2: > the structure of all of these definitions is an action the user might want to take

That was intentional. Is it a good or bad thing? 😅 

+1 on an example being helpful.

Reviewer 2: Ah, so the example prompt was very helpful! Turns out using this is not very fun, which I hadn’t tested 😅 

Basically the types are quite strict and there’s none of the helpful massaging that you get if you’re setting a sidebar in the Starlight config.

An example:

```astro
<StarlightPage
	frontmatter={{ title: 'Demo Page' }}
	sidebar={[
		{
			type: 'link',
			href: '/',
			label: 'Home',
			isCurrent: false,
			badge: undefined,
			attrs: {},
		},
		{
			type: 'group',
			label: 'Pages',
			entries: [
				{
					type: 'link',
					href: '/demo/',
					label: 'Demo',
					isCurrent: true,
					badge: undefined,
					attrs: {},
				},
			],
			badge: undefined,
			collapsed: false,
		},
	]}
/>
```

There’s a lot of boilerplate here, it could probably be more like this with some nice defaults:

```astro
<StarlightPage
	frontmatter={{ title: 'Demo Page' }}
	sidebar={[
		{ href: '/', label: 'Home' },
		{
			label: 'Pages',
			entries: [
				{ href: '/demo/', label: 'Demo', isCurrent: true },
			],
		},
	]}
/>
```

Wondering if we could just hold off documenting `sidebar` and ship a follow-up release that adds a nicer API and we document it then

Reviewer 2: Pulled the trigger on that — will follow up in a future PR.

Author: Random thoughts: If we don't like the API, wouldn't it be better to do the change now rather than later and force all plugins and `<StarlightPage/>` users to manually do the change? This will definitely be used if released as-is or people will ask for it considering all the broken plugins or previous `<Page>` users like Biome were specifying custom sidebars and it's been broken since October. I'm not sure if it's a good idea to release it as-is after all this time and then change it to an entirely different API in a few PRs.

If not set, the page will use the default global sidebar.

##### `hasSidebar`

**type:** `boolean`  
**default:** `false` if [`frontmatter.template`](/reference/frontmatter/#template) is `'splash'`, otherwise `true`

> Set if the sidebar should be displayed on this page.

Reviewer 3: ```suggestion
Whether or not the sidebar should be displayed on this page.
```

Similarly, as a more concrete example, this would be another way to frame all these options. Not a deal breaker, just pointing out that these rely on the type and default to do all the defining here, vs explainy text. Simply observing that it stood out to me!

Reviewer 2: ```suggestion
Control whether or not the sidebar should be displayed on this page.
```

##### `headings`

**type:** `{ depth: number; slug: string; text: string }[]`  
**default:** `[]`

Provide an array of all the headings on this page.
Starlight will generate the page table of contents from these headings if provided.

##### `dir`

**type:** `'ltr' | 'rtl'`  
**default:** the writing direction for the current locale

Set the writing direction for this page’s content.

##### `lang`

**type:** `string`  
**default:** the language of the current locale

Set the BCP-47 language tag for this page’s content, e.g. `en`, `zh-CN`, or `pt-BR`.

##### `isFallback`

**type:** `boolean`  
**default:** `false`

> Indicate if this page is untranslated in the current language and using [fallback content](/guides/i18n/#fallback-content).

Reviewer 3: ```suggestion
Indicate if this page is using [fallback content](/guides/i18n/#fallback-content) because there is no translation for the current language.
```
I feel like the important part here is that it's using fallback content (and the untranslated bit is secondary). So I might do something like this.
